applications/vanilla/modules/class.promotedcontentmodule.php

```diff
@@ -593,32 +602,59 @@ protected function Prepare(&$Content) {
 
       foreach ($Content as &$ContentItem) {
          $ContentType = val('RecordType', $ContentItem);
-
+         $UserID = GetValue('InsertUserID', $ContentItem);
          $Replacement = array();
-         $Fields = array('DiscussionID', 'CategoryID', 'DateInserted', 'DateUpdated', 'InsertUserID', 'Body', 'Format', 'RecordType');
+         $Fields = array('DiscussionID', 'DateInserted', 'DateUpdated', 'Body', 'Format', 'RecordType', 'Url', 'CategoryID', 'CategoryName', 'CategoryUrl', );
 
          switch (strtolower($ContentType)) {
             case 'comment':
                $Fields = array_merge($Fields, array('CommentID'));
 
                // Comment specific
-               $Replacement['Name'] = GetValueR('Discussion.Name', $ContentItem, val('Name', $ContentItem));
+               $Replacement['Name'] = sprintf(T('Re: %s'), GetValueR('Discussion.Name', $ContentItem, val('Name', $ContentItem)));
+               $url = CommentUrl($ContentItem);
                break;
 
             case 'discussion':
                $Fields = array_merge($Fields, array('Name', 'Type'));
+               $url = DiscussionUrl($ContentItem);
                break;
          }
 
+         $ContentItem['Url'] = $url;
+         if ($categoryId = val('CategoryID', $ContentItem)) {
+            $Category = CategoryModel::Categories($categoryId);
+            $ContentItem['CategoryName'] = val('Name', $Category);
+            $ContentItem['CategoryUrl'] = CategoryUrl($Category);
+         }
          $Fields = array_fill_keys($Fields, TRUE);
          $Common = array_intersect_key($ContentItem, $Fields);
          $Replacement = array_merge($Replacement, $Common);
          $ContentItem = $Replacement;
 
+
+
+         $UserFields = array('UserID', 'Name', 'Title', 'Location', 'PhotoUrl', 'RankName', '_CssClass', 'Url', 'Roles', 'RoleNames');
+
          // Attach User
-         $UserID = GetValue('InsertUserID', $ContentItem);
          $User = Gdn::UserModel()->GetID($UserID);
-         $ContentItem['Author'] = $User;
+         $RoleModel = new RoleModel();
+         $Roles = $RoleModel->GetByUserID($UserID)->ResultArray();
+         $RoleNames = '';
+         foreach($Roles as $Role) {
+            $RoleNames[] = val('Name', $Role);
+         }
+         $Replacement = array(
+            'Url' => Url(UserUrl($User), true),
+            'PhotoUrl' => UserPhotoUrl($User),
+            'RankName' => val('Name', RankModel::Ranks(val('RankID', $User)), null),
+            'RoleNames' => $RoleNames
+         );
+         $User = (array) $User;
+         $UserFields = array_fill_keys($UserFields, TRUE);
+         $Common = array_intersect_key($User, $UserFields);
+         $Replacement = array_merge($Common, $Replacement);
```

Comment:
These variable names aren't the best. Variable names should have guessable meaning (or at least be able to get in the ballpark) even if that's the only line of code you can see. My off-the-cuff suggestions:

* $UserFields => $AllowedUserFields
* $Common => $FilteredUser
* $Replacement (1st) => $UserAppend (?)
* $Replacement (2nd) => $User